STM32/FLASH_readwrite/Src/FLASH_readwrite.c

```diff
@@ -11,60 +12,136 @@
+{
+    uint32_t sector_start = getFlashSector(address);
+    uint32_t sector_end = getFlashSector(address + size);
+
+    // Ensure that the start and end addresses belong to the same FLASH sector.
+    if (sector_start != sector_end)
+    {
+        return 1;
+    }
+    return 0;
+}
+
+/*!
+** @brief Checks whether the user is trying to write to a memory range containing program memory.
+** @return Returns 1 if the memory range contains program memory, otherwise 0.
+*/
+static int isProgramMemory(uint32_t address, uint32_t size)
+{
+    if (address >= PROGRAM_START_ADDR && address < PROGRAM_END_ADDR)
+    {
+        return 1;
+    }
+
+    if ((address + size) >= PROGRAM_START_ADDR && (address + size) < PROGRAM_END_ADDR)
+    {
+        return 1;
+    }
+    return 0;
+}
+
+/***************************************************************************************************
+** PUBLIC FUNCTIONS
+***************************************************************************************************/
+
 /*!
 ** @brief Store data in FLASH memory.
+** @return Returns 1 on error and 0 on success.
 ** @note Does not store a CRC so data integrity verification is not possible when reading the 
 **       data at later stage.
 */
-void writeToFlash(uint32_t flash_address, uint8_t *data, uint32_t size)
+int writeToFlash(uint32_t flash_address, uint8_t *data, uint32_t size)
 {
     __HAL_RCC_WWDG_CLK_DISABLE();
+
+    // User is trying to write to an address containing program data
+    if (isProgramMemory(flash_address, size)) { return 1; }
+
     uint32_t flash_sector = getFlashSector(flash_address);
+
+    // User is trying to write to non-valid sector
+    if (flash_sector == 10) { return 1; }
+
+    // User is trying to write to more than one sector
+    if (isWriteWithinSector(flash_address, size) != 0){ return 1; }
+
+    // If !HAL_OK FLASH unlock failed and cannot be erased or written to
+    if (HAL_FLASH_Unlock() != HAL_OK) { return 1; }
+
     // Erase the sector before write
-    HAL_FLASH_Unlock();
     FLASH_Erase_Sector(flash_sector, FLASH_VOLTAGE_RANGE_3);
     HAL_FLASH_Lock();
 
     // Write to sector
     HAL_FLASH_Unlock();
-
+    // Store data
     for(uint32_t i=0; i<size; i++)
     {
-        HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, flash_address+i , data[i]);
+        if (HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, flash_address+i , data[i]) != HAL_OK) { return 1; }
     }
+
     HAL_FLASH_Lock();
     __HAL_RCC_WWDG_CLK_ENABLE();
+    return 0;
 }
 
 /*!
 ** @brief Read data from FLASH memory at flash_adress.
 */
-void readFromFlash(uint32_t flash_address, uint8_t *data, uint32_t size)
+int readFromFlash(uint32_t flash_address, uint8_t *data, uint32_t size)
 {
-    for(uint32_t i=0; i<size; i++)
-    {
-        *(data + i) = *( (uint8_t *)(flash_address+i) );
-    }
+    // User is trying to read from an address containing program data
+    if (isProgramMemory(flash_address, size)) { return 1; }
```

Comment:
But this should be allowed I think? There is nothing dangerous about that

Reply:
Fair I will remove those guards